be/src/exec/spill/spilled_stream.cpp

```diff
@@ -36,16 +36,263 @@
+        _chunk_buffer.put(std::move(res.value()));
+        return Status::OK();
+    } else if (res.status().is_end_of_file()) {
+        mark_is_eof();
+        return Status::OK();
+    } else {
+        return res.status();
+    }
+}
+
+class SortedFileStream final : public SpilledInputStream {
+public:
+    SortedFileStream(SpilledInputStreamList streams, RuntimeState* state);
+    ~SortedFileStream() override = default;
+
+    Status init(const SortExecExprs* exprs, const SortDescs* descs);
+    StatusOr<ChunkUniquePtr> read(SpillFormatContext& context) override;
+    bool is_ready() override { return _merger.is_data_ready(); }
+    void close() override {}
+    std::vector<std::shared_ptr<BufferedSpilledStream>> raw_streams() { return _buffered_streams; }
+
+private:
+    std::atomic<bool> _eos{};
+    SpillFormatContext _read_context;
+    Status _status;
+
+    std::vector<std::shared_ptr<BufferedSpilledStream>> _buffered_streams;
+
+    SpilledInputStreamList _streams;
+    CascadeChunkMerger _merger;
+};
+
+SortedFileStream::SortedFileStream(SpilledInputStreamList streams, RuntimeState* state)
+        : _streams(std::move(streams)), _merger(state) {}
+
+Status SortedFileStream::init(const SortExecExprs* sort_exprs, const SortDescs* descs) {
+    std::vector<ChunkProvider> providers;
+    for (auto& stream : _streams) {
+        _buffered_streams.emplace_back(std::make_unique<BufferedSpilledStream>(chunk_buffer_max_size, stream));
+        auto buffered_stream = _buffered_streams.back();
+        auto chunk_provider = [buffered_stream, this](ChunkUniquePtr* out_chunk, bool* eos) {
+            if (out_chunk == nullptr || eos == nullptr) {
+                return buffered_stream->has_chunk();
+            }
+
+            if (!buffered_stream->has_chunk()) {
+                return false;
+            }
+
+            auto stv = buffered_stream->read_from_buffer();
+            // auto stv = buffered_stream->read(_read_context);
+            if (!stv.status().ok()) {
+                if (!stv.status().is_end_of_file()) {
+                    _status.update(stv.status());
+                }
+                buffered_stream->mark_is_eof();
+                *eos = true;
+                return false;
+            }
+
+            *out_chunk = std::move(stv.value());
+            return true;
+        };
+        providers.emplace_back(std::move(chunk_provider));
+    }
+    RETURN_IF_ERROR(_merger.init(providers, &(sort_exprs->lhs_ordering_expr_ctxs()), *descs));
+    return Status::OK();
+}
+
+StatusOr<ChunkUniquePtr> SortedFileStream::read(SpillFormatContext& context) {
+    ChunkUniquePtr chunk;
+    bool should_exit = false;
+    RETURN_IF_ERROR(_merger.get_next(&chunk, &_eos, &should_exit));
+
+    if (chunk && !chunk->is_empty()) {
+        return std::move(chunk);
+    }
+
+    if (_eos) {
+        return Status::EndOfFile("eos");
+    }
+
+    // data is not arrived
+    DCHECK(should_exit);
+    return std::make_unique<Chunk>();
```

Comment:
is it better to return nullptr if no data?

Reply:
return empty chunk is more safely